src/TwilioProvider.php

```diff
@@ -22,19 +24,19 @@ public function boot()
                 );
             });
 
-        $this->app->bind(TwilioService::class, function () {
-            $config = $this->app['config']['services.twilio'];
-            $account_sid = Arr::get($config, 'account_sid');
-            $username = Arr::get($config, 'username');
-            if (! empty($username)) {
-                $password = Arr::get($config, 'password');
+        $this->app->bind(TwilioService::class, function (Application $app) {
+            /** @var TwilioConfig $config */
+            $config = $app->make(TwilioConfig::class);
 
-                return new TwilioService($username, $password, $account_sid);
-            } else {
-                $auth_token = Arr::get($config, 'auth_token');
+            if ($config->usingUsernamePasswordAuth()) {
+                return new TwilioService($config->getUsername(), $config->getPassword(), $config->getAccountSid());
+            }
 
-                return new TwilioService($account_sid, $auth_token);
+            if ($config->usingTokenAuth()) {
+                return new TwilioService($config->getAccountSid(), $config->getAuthToken());
             }
+
+            throw InvalidConfigException::missingConfig();
         });
```

Comment:
I would also rework the binding for the `TwilioChannel`.
At the present state is looks like following:

```php
$this->app->when(TwilioChannel::class)
            ->needs(Twilio::class)
            ->give(function () {
                return new Twilio(
                    $this->app->make(TwilioService::class),
                    $this->app->make(TwilioConfig::class)
                );
            });
```

It seems in this way a new instance of `Twilio` will be instantiated per each notification using `TwilioChannel`, which is redundant.

Reply:
Have switched to a singleton